pulsar-broker/src/main/java/org/apache/pulsar/broker/service/TopicEventsListener.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.broker.service;
+
+public interface TopicEventsListener {
+
+    enum TopicEvent {
+        // create events included into load events
+        BEFORE_TOPIC_CREATE,
+        BEFORE_TOPIC_LOAD,
+        BEFORE_TOPIC_UNLOAD,
+        BEFORE_TOPIC_DELETE,
+
+        // after and failed events are mutually exclusive
+        AFTER_TOPIC_CREATE,
+        AFTER_TOPIC_LOAD,
+        AFTER_TOPIC_UNLOAD,
+        AFTER_TOPIC_DELETE,
+
+        FAILED_TOPIC_CREATE,
+        FAILED_TOPIC_LOAD,
+        FAILED_TOPIC_UNLOAD,
+        FAILED_TOPIC_DELETE
+    }
+
+    void handleEvent(String topicName, TopicEvent event);
```

Comment:
I added throwable
Topic can be accessed via brokerService.getTopics().get(topicName)

Reply:
In case of unload() 
Can I still  access the instance of Topic ?

Also, what is the status of the Topic? Is it still possible to access internal components, especially the ManagedLedger object?

Reply:
In case of the  load event, is the Topic already active/accessible? 
We should state that in the javadocs and with a test case

Reply:
> In case of the load event, is the Topic already active/accessible?

It should be accessible after the load / managed ledger open

> In case of unload()
> Can I still access the instance of Topic ?

not after unload. Isn't topic name enough to locate it in protocol handler's internal caches?
before load currently also fires after the topic is closed / on removal from BrokerService's cache. 
If needed, it can be moved into topic's close() with some extra work. The question is do you really need anything but the topic name in that case.

Reply:
I am fine with the current approach as long as it is not possible that a new Topic instance is started and registered in the internal map before the listener is executed.

During some error scenarios it is possible le that the broker loads a topic, then unloads it and then loads it again.
I think that all of those actions should run sequentially so maybe there is no problem.

Can you please double check as well?

Reply:
I think there is a risk of load/unload events mixing up;
Unload is tracked on removal from cache / after it succeeds, load/create happens in computeIfAbsent and there is no other locking. We can consider adding locks there at expense of perf, if needed. I'll look a bit more to check if there is anything else that guards load/unload from racing against each other. 

E.g. removeTopicFromCache can remove topic from cache and before it runs `compactor.getStats().removeTopic` the load can load the topic again, the events affected by this too.